README.md
#jac [![Build Status](https://travis-ci.org/busbud/jac.png)](https://travis-ci.org/busbud/jac)

> Really when you push something out live to the world, you never want to change it without changing the name because
> there are so many misconfigured proxies out there. About 1% to 10% of your users will never get an update
> unless you change the name.

> You can make it cachable for 10 years, you're never going to push a change without changing the name of the file.<br/>

> <cite>- Steve Souders. HTML5DevConf (Jan 10, 2013).
  <a href="http://marakana.com/s/post/1360/cache_is_king_steve_souders_html5_video">Cache is king</a>
  availalable from <a href="http://mrkn.co/3wzua">http://mrkn.co/3wzua</a></cite>


jac provides methods to reference asset urls using permanently cachable urls.

It achieves this by using urls unique to each version of a given asset, and provides middleware to serve the asset with
long cache durations.

#Usage
jac middleware will handle asset file serving and asset url resolution

Install it using

    npm install jac

## Example
The following example shows how to wire up jac with an express 3.x app
 * App code sets up the jac middleware
 * A JSON file stores the configuration for the jac middleware
 * A script updates the configuration file

### App
```js
var express = require('express')
  , jac     = require('jac')
  , app     = express()
  , config  = require('./config');

// Add middleware for all requests
app.use(jac.middleware(config));

// Add view that resolves an image url
app.get('/someview', function (req, res) {
  var jac = res.locals.jac        // returns jac view helper
    , key = '/images/happy.png'     // matches config key
    , url = jac.resolve(key);       // returns url with digest, handled by middleware

  res.send(url);
});
```

### Config
The configuration file is updated using jac's digester and read by the app. Here's a sample config json file loaded
by the app.

```js
{
  // required - files served by jac
  assets: [{
    fullPath: require('path').resolve(__dirname, './public/images/happy.png'),  // local file path
    key:      '/images/happy.png',                                              // key used in views: jac.resolve('/images/happy.png')
    route:    '/images/b64Digest/happy.png',                                    // route for middleware
    url:      '//cdn.host.net/images/b64Digest/happy.png'                       // url output to response and css
  }],

  // optional - cache control max age in seconds (default 2 weeks)
  maxAge: 5443200
}
```

### Update script
The update script runs the digester against the files specified in configuration and updates the `files` key in the
configuration file. This script can be run from `package.json`'s scripts section after install.

```bash
jac --config ./config.json
```

Alternatively, the digester can be wired up manually

```js
var path   = './config'
  , config = require(path)
  , digest = require('jac').digest;

function save (err, assets) {
  if (err) throw err;

  config.assets = assets;
  fs.writeFile(path, JSON.stringify(config, null, 2));
}

digest(config, save);
```

### Replacing references in CSS
jac can be configured to process CSS files as part of the update process. It will ignore image references from external
sites (eg urls with an [authority](http://medialize.github.com/URI.js/docs.html#accessors-authority)) and data-uris,
and will attempt to resolve all other urls.

If it fails to resolve a url, it will throw an error, allowing you to find the problematic reference. Most likely, this
will easily be corrected by adjusting the path to the image to make it root relative.

Here's an example that will result in jac replacing the image reference

__src/stylesheets/main.css__

```css
body {background: url(/images/happy.png);}
```

__config.json__

```js
{
  // required - files served by jac
  assets: [{
    fullPath: require('path').resolve(__dirname, './public/images/happy.png'),  // local file path
    key:      '/images/happy.png',                                              // key used in views: jac.resolve('/images/happy.png')
    route:    '/images/b64Digest/happy.png',                                    // route for middleware
    url:      '//cdn.host.net/images/b64Digest/happy.png'                       // url output to response and css
  }],

  css: {
    'public/stylesheets/main.css': 'src/stylesheets/main.css'                   // format <output>: <input>
  }
}
```

To run the CSS replacement, update the jac config file to include the css property and use the following command

```bash
css --config ./config.json
```

The file at public/stylesheets/main.css will now contain the following

```css
body {background: url(//cdn.host.net/images/b64Digest/happy.png);}
```


## Compatibility
This version of jac is compatible with express 3.x.
It depends on `res.locals` to get and set the view local `jac` via middleware.

## Production
To support production usage with the best performance, it is required that a pre-processing step be executed to
generate the image file digests stored in configuration.

```bash
jac --config ./config.json
```

It is recommended that the config file already exist, and the config file used by the app
always be updated before deploying to production. This can be automated using `package.json`'s
scripts section after install.

```js
{
  "name": "myproject",
  "dependencies": {
    "jac": "latest"
  },
  "scripts": {
    "postinstall": "./node_modules/jac/bin/jac"
  }
}
```

By default, jac will load the config from the `jac.json` file at the project root.

# Running Tests
To run the test suite first invoke the following command within the repo, installing the development dependencies:

    npm install

then run the tests:

    npm test

# License

(The MIT License)

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the 'Software'), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

CHANGELOG.md
## v0.1.0

* Update for express 3 compatibility.

## v0.0.5

* Add 'Vary: Accept-Encoding' response header to handle proxies and CDNs

## v0.0.4

* Add support to use `jac.resolve` with CSS files (offline)
* Allow json and js files to be used as configuration inputs for ./bin/jac

## v0.0.3

* Remove mtime in digest

## v0.0.2

* Generalize to any static asset
* Switch to use `jac.resolve` instead of `jac.img.resolve` and

## v0.0.1

* initial version


test/css.js
"use strict";

var should = require('should');
var fs     = require('fs');
var path   = require('path');
var async  = require('async');
var _      = require('lodash');

describe('CSS filter', function () {
  var css = require('../src/css');

  /**
   * Verify the substitution process
   *
   * @param {Object} config
   * @param {Object} files, where key=output, value=input
   * @param outputPath
   * @param expectedPath
   * @param done
   */
  function processAndVerify(config, files, outputPath, expectedPath, done) {
    config     = _.defaults(config, {silent:true});
    config.css = files;

    css.process(config, verify);

    function verify() {
      var outputContent   = fs.readFileSync(outputPath).toString();
      var expectedContent = fs.readFileSync(expectedPath).toString();

      outputContent.should.equal(expectedContent);
      done();
    }
  }


  describe('process', function () {
    var config ={
      assets: [
        {
          fullPath: path.resolve(__dirname, './fixtures/spacer.gif'),
          key: '/images/spacer.gif',
          route: '/images/b64Digest/spacer.gif',
          url: '/images/b64Digest/spacer.gif'
        }
      ]
    };

    it('should have a method called process', function () {
      should.exist(css);
      css.should.have.ownProperty('process');
      css.process.should.be.instanceof(Function);
    });

    it('should replace static image references', function (done) {
      var expected = path.resolve(__dirname, './fixtures/imagereferences.expected.css');
      var input    = path.resolve(__dirname, './fixtures/imagereferences.css');
      var output   = path.resolve(__dirname, './fixtures/temp/imagereferences.out.css');
      var files    = {};

      // Output is sourced from input
      files[output] = input;

      processAndVerify(config, files, output, expected, done);
    });

    it('should ignore previously resolved static image references', function (done) {
      var expected = path.resolve(__dirname, './fixtures/imagereferences.expected.css');
      var input    = expected;
      var output   = path.resolve(__dirname, './fixtures/temp/imagereferences.out.css');
      var files    = {};

      // Output is sourced from input
      files[output] = input;

      processAndVerify(config, files, output, expected, done);
    });
  });

  describe('process with host name', function () {
    var config ={
      host: "cdn.net",
      assets: [
        {
          fullPath: path.resolve(__dirname, './fixtures/spacer.gif'),
          key: '/images/spacer.gif',
          route: '/images/b64Digest/spacer.gif',
          url: '//cdn.net/images/b64Digest/spacer.gif'
        }
      ]
    };

    it('should replace static image references (and include host name)', function (done) {
      var expected = path.resolve(__dirname, './fixtures/imagereferences.withhost.expected.css');
      var input    = path.resolve(__dirname, './fixtures/imagereferences.css');
      var output   = path.resolve(__dirname, './fixtures/temp/imagereferences.withost.out.css');
      var files    = {};

      // Output is sourced from input
      files[output] = input;

      processAndVerify(config, files, output, expected, done);
    });

    it('should ignore previously resolved static image references, even with host name', function (done) {
      var expected = path.resolve(__dirname, './fixtures/imagereferences.withhost.expected.css');
      var input    = expected;
      var output   = path.resolve(__dirname, './fixtures/temp/imagereferences.withost.out.css');
      var files    = {};

      // Output is sourced from input
      files[output] = input;

      processAndVerify(config, files, output, expected, done);
    });
  });
});

test/digester.js
"use strict";

var should = require('should');
var path   = require('path');

describe('Digester', function () {
  var digester = require('../src/digester');

  describe('digest directory', function () {
    var entries, error;

    before(function (done) {
      var opts = {
        root:       __dirname,
        fileFilter: ["*.gif"],
        silent:     true
      };

      digester.process(opts, cb);

      function cb (err, res) {
        error = err;
        entries = res;
        done();
      }
    });

    it('entries should be an array', function () {
      entries.should.be.an.instanceOf(Array);
    });

    it('entries should be .gif', function () {
      entries.forEach(isGif);

      function isGif(e) {
        path.extname(e.fullPath).should.equal(".gif");
      }
    });

    it('entries should sorted by fullPath', function () {
      var paths = entries.map(toPath).sort();

      entries.forEach(checkSort);

      function toPath(e) {
        return e.fullPath;
      }

      function checkSort(e, i, arr) {
        e.fullPath.should.equal(paths[i]);
      }
    });

    it('entries should have properties', function () {
      entries.forEach(verify);

      function verify (e) {
        e.should.have.property('fullPath');
        e.should.have.property('key');
        e.should.have.property('url');
        e.should.have.property('digest');
        e.should.have.property('route');
      }
    });

    /**
      Don't include a query string in the URL for static resources.

      Most proxies, most notably Squid up through version 3.0, do not cache resources with a "?" in
      their URL even if a Cache-control: public header is present in the response. To enable proxy
      caching for these resources, remove query strings from references to static resources, and
      instead encode the parameters into the file names themselves.

      - https://developers.google.com/speed/docs/best-practices/caching

     */
    it('entries\' route should not have a querystring, but should contain digest in path', function () {
      entries.forEach(verify);

      function verify (e) {
        e.route.indexOf('?').should.equal(-1);
        e.route.indexOf(e.digest).should.not.equal(-1);
        e.url.indexOf('?').should.equal(-1);
        e.url.indexOf(e.digest).should.not.equal(-1);
      }
    });

    it('entries\' key should start with a \'/\'', function () {
      entries.forEach(verify);

      function verify (e) {
        e.key[0].should.equal('/');
      }
    });
  });

  describe('digest directory with explicit host', function () {
    var entries, error;

    before(function (done) {
      var opts = {
        root:       __dirname,
        fileFilter: ["*.gif"],
        silent:     false,
        host:       'cdn.net'
      };

      digester.process(opts, cb);

      function cb (err, res) {
        error = err;
        entries = res;
        done();
      }
    });

    it('entries should have the host set only on the url', function () {
      entries.forEach(verify);

      function verify (e) {
        e.should.have.property('fullPath');
        e.should.have.property('key');
        e.should.have.property('url');
        e.should.have.property('digest');
        e.should.have.property('route');

        var prefix = '//cdn.net/';
        e.url.indexOf(prefix).should.equal(0);
        e.route.indexOf(prefix).should.equal(-1);
        e.key.indexOf(prefix).should.equal(-1);
      }
    });

    it('entries\' route should not have a querystring, but should contain digest in path', function () {
      entries.forEach(verify);

      function verify (e) {
        e.route.indexOf('?').should.equal(-1);
        e.route.indexOf(e.digest).should.not.equal(-1);
        e.url.indexOf('?').should.equal(-1);
        e.url.indexOf(e.digest).should.not.equal(-1);
      }
    });
  });
});

test/middleware.js
"use strict";

var should = require('should');
var path   = require('path');
var assert = require('assert');
var express= require('express');
var request= require('supertest');

describe('Middleware', function () {
  var factory = require('../src/middleware');
  var twoweeks= 3600 * 24 * 14;

  describe('with single config entry', function () {
    var middleware
      , app;

    before(function () {
      middleware = factory.create({
        assets: [
          {
            fullPath: path.resolve(__dirname, './fixtures/spacer.gif'),
            key:      '/images/spacer.gif',
            route:    '/images/b64Digest/spacer.gif',
            url:      '/images/b64Digest/spacer.gif'
          }
        ]
      });

      app = express.createServer();

      app.use(middleware);

      // Set up views for testing middleware
      app.get('/view-spacer', function (req, res) {
        res.send(res.locals.jac.resolve('/images/spacer.gif'));
      });
      app.get('/view-unresolved', function (req, res) {
        res.send(res.locals.jac.resolve('/images/noexisto.gif'));
      });
      app.use(function (err, req, res, next) {
        res.send(err.message, 500);
      });
    });

    after(function () {
      app = null;
    });

    it('should not be null', function () {
      should.exist(middleware);
    });

    it('should be a function', function () {
      middleware.should.be.instanceof(Function);
    });

    it('should stream the img file on exact url match', function (done) {
      request(app)
        .get('/images/b64Digest/spacer.gif')
        .expect('Cache-Control', 'public, max-age=' + twoweeks)
        .expect('Content-Type', 'image/gif')
        .expect('Content-Length', '43')
        .expect('Vary', 'Accept-Encoding')
        .expect(200, done);
    });

    it('should 404 on digest mismatch', function (done) {
      request(app)
        .get('/images/b64DigestX/spacer.gif')
        .expect(404, done);
    });

    it('should 404 on digest omission', function (done) {
      request(app)
        .get('/images/spacer.gif')
        .expect(404, done);
    });

    it('should resolve the img url for other responses', function (done) {
      request(app)
        .get('/view-spacer')
        .expect(200, '/images/b64Digest/spacer.gif', done);
    });

    it('should throw an error for unknown img urls', function (done) {
      request(app)
        .get('/view-unresolved')
        .expect(500, 'jac: key /images/noexisto.gif not found, regenerate jac config or update key value to match key in jac config', done);
    });
  });
});

src/css.js
"use strict";

var mkdirp     = require('mkdirp');
var URIjs      = require('URIjs');
var async      = require('async');
var path       = require('path');
var fs         = require('fs');
var _          = require('lodash');
var middleware = require('./middleware');

var extractor  = /\burl\(['"]?([^\)]+?)['"]?\)/gi;

var css = module.exports = {};

/**
 * Replaces all url references to images in CSS files with
 * cachable url references
 *
 * @param {Object} config, with a key called 'css' with the value {dest1:src1, dest2:src2}
 * @param {Function} done
 */
css.process = function (config, done) {
  var files       = config.css;
  var silent      = config.silent;
  var concurrency = 1;
  var sources     = _.values(files);
  var sourcemap   = _.invert(files);

  var assetsByUrl = config.assets.reduce(function (m, asset) {
    m[asset.url] = asset;
    return m;
  }, {});

  var assetsByKey = config.assets.reduce(function (m, asset) {
    m[asset.key] = asset;
    return m;
  }, {});

  async.forEachLimit(
    sources,
    concurrency,
    replace,
    done
  );

  function log () {
    if (!silent) {
      console.log.apply(null, Array.prototype.slice.call(arguments));
    }
  }

  function replace(file, done) {
    var steps = [];

    steps.push(function outputdir(cb) {
      var filepath = sourcemap[file];
      var dir      = path.dirname(filepath);
      var exists   = fs.exists || path.exists;

      exists(dir, function (flag) {
        if (!flag) {
          mkdirp(dir, function (err) {
            cb(err);
          });
        } else {
          cb(null);
        }
      });
    });

    steps.push(function read(cb) {
      fs.readFile(file, cb);
    });

    steps.push(function makeCachable(contents, cb) {
      var count = 0;
      var replaced = contents.toString().replace(extractor, function (prop, key) {
        var uri = new URIjs(key);

        if (!uri.authority() && !uri.protocol()) {
          count++;
          if (key in assetsByUrl) {
            // Route was already transformed
            return prop;
          }

          var asset = assetsByKey[key];
          if (asset) {
            return prop.replace(key, asset.url);
          }
          else {
            throw new Error('jac: file:' + file + ': key ' + key + ' not found, regenerate jac config (or refresh/update your css)');
          }
        } else {
          return prop;
        }
      });

      cb(null, replaced, count);
    });

    steps.push(function write(contents, count, cb) {
      var output = sourcemap[file];
      log('updating %s with %d replacement(s)', output, count);

      fs.writeFile(output, contents, cb);
    });

    async.waterfall(steps, done);
  }
};

src/digester.js
"use strict";

var _        = require('lodash');
var url      = require('url');
var path     = require('path');
var async    = require('async');
var URIjs    = require('URIjs');
var readdirp = require('readdirp');

var digester = module.exports = {};

function selectStrategy(opts) {
  var strategy = opts.strategy;
  var digestfn = strategy;

  if (!_.isFunction(strategy)) {
    strategy = require('./strategy/hash').create(opts);
    digestfn = strategy.digest.bind(strategy);
  }

  return digestfn;
}

/**
 * Processes all files under opts.root and generates the config
 * entries for each matched file.
 *
 * @param opts     {root, vdir, strategy, strategy-config}
 * @param callback fn(err, entries)
 */
digester.process = function (opts, callback) {
  var root = opts.root || '.';
  var vdir = opts.vdir || '/';
  var host = opts.host || '';
  var filter   = opts.fileFilter || ['*.gif', '*.jpg', '*.jpeg', '*.png'];
  var silent   = opts.silent;
  var strategy = selectStrategy(opts);
  var start    = Date.now();
  var base     = process.cwd();

  vdir = url.parse(vdir);

  function log () {
    if (!silent) {
      console.log.apply(null, Array.prototype.slice.call(arguments));
    }
  }

  readdirp({root: root, fileFilter: filter}, function (err, tree) {
    if (err) {
      return callback(err);
    }

    var processed = [];
    var entries = tree.files.filter(notHidden).map(toPaths);
    log('processing %d file(s)', entries.length);

    async.forEachLimit(
      entries,
      10,
      processEntry,
      complete
    );

    function notHidden(f) {
      return path.basename(f.path)[0] !== '.';
    }

    function toPaths (f) {
      return {
        fullPath: path.relative(base, f.fullPath),
        key: url.resolve(vdir, f.path),
        url: (new URIjs(url.resolve(vdir, f.path))).host(host).href()
      };
    }

    function processEntry(entry, done) {
      log('   \x1b[36mprocess\x1b[0m : ' + entry.key);

      strategy(entry, function (err, entry) {
        if (!err) {
          processed.push(entry);
        }
        done(err);
      });
    }

    function complete (err) {
      if (!err) {
        var config = _(processed)
          .sortBy(function (entry) {
            return entry.fullPath;
          })
          .map(function (entry) {
            // generate the urls and route by injecting the digest into the path
            var original = new URIjs(entry.url);
            var modified = original
              .clone()
              .directory(original.directory() + '/' + entry.digest);

            entry.url   = modified.href();
            entry.route = modified.host('').href();

            return entry;
          })
          .value();

        var delay = Date.now() - start;
        log('processed %d file(s) in %d ms', entries.length, delay);

        return callback(null, config);
      }

      callback(err);
    }
  });
};

src/strategy/hash.js
"use strict";

var fs     = require('fs');
var crypto = require('crypto');

module.exports.create = function (opts) {
  return new HashStrategy(opts);
};

function HashStrategy(opts) {
  this.algorithm = opts && opts.algorithm || 'md5';
  this.salt      = opts && opts.salt || '';
  this.length    = opts && opts.length || 7;
}

/**
 * Generates and sets an entry's digest
 *
 * @param entry {fullpath, key, url, mtime}
 * @param done  fn(err, entry)
 */
HashStrategy.prototype.digest = function (entry, done) {
  var self = this;
  var hash = crypto.createHash(self.algorithm);

  var s = fs.ReadStream(entry.fullPath);

  hash.update(self.salt.toString());

  s.on('data', function(d) {
    hash.update(d);
  });

  s.on('error', function (err) {
    done(err);
  });

  s.on('end', function() {
    var d = hash.digest('base64');

    // Ensure url-safe base64
    entry.digest = d.substr(0, self.length)
      .replace(/\+/g, "-")
      .replace(/\//g, "_");

    done(null, entry);
  });
};


src/middleware.js
"use strict";

var _    = require('lodash');
var send = require('send');

/**
 * Creates jac-img middleware for specified config
 *
 * @param config
 * @return {Function}
 */
module.exports.create = function (config) {
  var assets = config.assets;

  var assetsByRoute = assets.reduce(function (memo, entry) {
    memo[entry.route] = entry;
    return memo;
  }, {});

  var assetsByKey = assets.reduce(function (memo, entry) {
    memo[entry.key] = entry;
    return memo;
  }, {});

  var twoweeks  = 14 * 24 * 3600;
  config.maxAge = (config.maxAge || twoweeks) * 1000; // in milliseconds, for `send`

  /**
   * Returns the route url with digest for the resource specified by the key
   *
   * @param key
   * @return {String}
   * @throws {Error} if key not found
   */
  function resolveAsset (key) {
    var asset = assetsByKey[key];

    if (!asset) {
      throw new Error('jac: key ' + key + ' not found, regenerate jac config or update key value to match key in jac config');
    }

    return asset.url;
  }

  function middleware (req, res, next) {
    var asset = assetsByRoute[req.url];

    if (!asset) {

      var jac = res.locals.jac || {};
      jac.resolve = resolveAsset;
      res.locals.jac = jac;
      
      return next();
    }

    // Add route information for logging purposes
    req.route = {
      path: '<jac.middleware>'
    };

    // Add Vary: Accept-Encoding to deal with proxies on CDNs and such
    res.setHeader('Vary', 'Accept-Encoding');

    send(req, asset.fullPath)
      .maxage(config.maxAge)
      .on('error', next)
      .pipe(res);
  }

  return middleware;
};